### ISSUE TITLE
Merging unstable into form-components breaks editor/graph synchronization

### ISSUE THREAD

**Issue by fdurupinar** [PathwayCommons/factoid#295]
## Environment info

- Browser version (client) : Chrome  66.0.3359.181 (Official Build) (64-bit)
- Node.js version (server) :v8.2.1

## Description 

Synchronization between the graph and form editors is broken after merging unstable into form-components.

I.e. commit 7a391a98b8314b3c128c1211c301818fb02272fe was working fine, but  1f9be760ef48ac0667646f2b47c25bd81586ead8 is buggy due to the recent updates.

## Current (buggy) behaviour
- Interactions do not appear correctly on the other side when added through graph or form.
- Completing an entity on the form side would automatically complete the entity on the graph side in the previous commit. This is not the case anymore.

**Comment by d2fong** [comment]
@fdurupinar Do you have the merged version of the branch somewhere?

**Comment by metincansiper** [comment]
@d2fong it is merged to form-components branch. Merging is done by 1f9be76 so 1f9be76 and later are buggy. The bug comes from [this line](https://github.com/PathwayCommons/factoid/commit/1f9be760ef48ac0667646f2b47c25bd81586ead8#diff-be250af46404c5fa3538c8fa311ac7e6R132). Some ``ele`` there are ``undefined``, so the following error is thrown:

"Unhandled rejection TypeError: Cannot read property 'completed' of undefined"

**Comment by metincansiper** [comment]
I suppose [the line](https://github.com/PathwayCommons/factoid/commit/1f9be760ef48ac0667646f2b47c25bd81586ead8#diff-be250af46404c5fa3538c8fa311ac7e6R132) I mentioned above was not the cause of problem itself but it was one of the effects of the problem. Browser shows that target of edge could not be created before that error:

"Can not create edge `031d1d45-d8af-470a-ade8-481692e3a8e9` with nonexistant target `260efa21-1781-490c-9e5e-81ddaa52e34d`"

**Comment by maxkfranz** [comment]
Does the error happen only once or many times per session?

**Comment by maxkfranz** [comment]
In the form branch, there was also changes to the model made.  If those changes are reverted, what is the effect on this?

**Comment by metincansiper** [comment]
>Does the error happen only once or many times per session?

It happens in network editor any time a new interaction is created from form editor.

**Comment by metincansiper** [comment]
@maxkfranz @d2fong @fdurupinar 
I found out the lines of codes that causes the bug or at least the bug disappears when I comment out that lines.

The lines I mention are 

```
this.props.document.on('add', this.onAdd);
this.props.document.on('remove', this.onRemove);
```

in task-list.js (There are 2 occurrence of each of the lines above in the file, one for TaskList one for TaskListButton). The bug dissappears when I tried to comment out 4 occurrences of the lines.

Is this something that must be fixed from form editor or is it a more general problem related to TaskList and TaskListButton?

**Comment by maxkfranz** [comment]
It sounds like an exception in the task list blocks some further ui/sync code from running.  I'll take a look tomorrow at the task list to see if anything in the listeners stands out.

**Comment by maxkfranz** [comment]
See the form-fixes branch: https://github.com/PathwayCommons/factoid/commits/form-fixes

I made several commits that fix the errors originating in the form.  All but the task list error are fixed by changing the form code only.  There are some other changes, in the model for example, that should help avoiding (or at least more easily finding) similar errors in future.

The three bugs resolved include:

- The 'can not create edge' bug that Metin mentioned
- The 'can not read completed' bug that Metin mentioned
- The invalid association data init bug.  I don't see anyone else mention this one in this ticket, but I reproduced it several times on the form-components branch.

It also seems like there is a fair amount of form code that could benefit from cleanup and refactoring.  Maybe we should create a separate issue for code review of the form.

**Comment by d2fong** [comment]
Can you confirm that the task list error is still not fixed? I am checking this branch on my local machine and it doesn't seem to be throwing any errors.

**Comment by metincansiper** [comment]
>Can you confirm that the task list error is still not fixed? I am checking this branch on my local machine and it doesn't seem to be throwing any errors.

Do you mean you cannot reproduce error in form-components? It was appearing in network editor when you are adding an interaction from the form editor in that branch? Seems like it is fixed in form-fixes.

**Comment by d2fong** [comment]
I am referring to this comment:

> All but the task list error are fixed by changing the form code only

**Comment by maxkfranz** [comment]
It should be working in the form-fixes branch.  It would be good to have other people verify on their machines, as they may have different rethinkdb, node, npm, etc.
